src/sentry/api/endpoints/organization_member_invite/index.py

```diff
@@ -0,0 +1,337 @@
+        assigned_org_role = request.data.get("orgRole") or organization_roles.get_default().id
+        allowed_roles = get_allowed_org_roles(request, organization, creating_org_invite=True)
+
+        # We allow requests from integration tokens to invite new members as the member role only
+        if not allowed_roles and request.access.is_integration_token:
+            # Error if the assigned role is not a member
+            if assigned_org_role != "member":
+                raise serializers.ValidationError(
+                    "Integration tokens are restricted to inviting new members with the member role only."
+                )
+            allowed_roles = [organization_roles.get("member")]
+
+        serializer = OrganizationMemberInviteRequestSerializer(
+            data=request.data,
+            context={
+                "organization": organization,
+                "allowed_roles": allowed_roles,
+                "allow_retired_roles": not features.has("organizations:team-roles", organization),
+            },
+        )
+
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=400)
+
+        result = serializer.validated_data
+
+        if ratelimits.for_organization_member_invite(
+            organization=organization,
+            email=result["email"],
+            user=request.user,
+            auth=request.auth,
+        ):
+            metrics.incr(
+                "member-invite.attempt",
+                instance="rate_limited",
+                skip_internal=True,
+                sample_rate=1.0,
+            )
+            return Response({"detail": ERR_RATE_LIMITED}, status=429)
+
+        is_member = not request.access.has_scope("member:admin") and (
+            request.access.has_scope("member:invite")
+        )
+        # if Open Team Membership is disabled and Member Invites are enabled, members can only invite members to teams they are in
+        members_can_only_invite_to_members_teams = (
+            not organization.flags.allow_joinleave and not organization.flags.disable_member_invite
+        )
+        has_teams = bool(result.get("teams"))
+
+        if is_member and members_can_only_invite_to_members_teams and has_teams:
+            requester_teams = set(
+                OrganizationMember.objects.filter(
+                    organization=organization,
+                    user_id=request.user.id,
+                    user_is_active=True,
+                ).values_list("teams__slug", flat=True)
+            )
+            team_slugs = [team.slug for team in result.get("teams", [])]
+            # ensure that the requester is a member of all teams they are trying to assign
+            if not requester_teams.issuperset(team_slugs):
+                return Response(
+                    {"detail": "You cannot assign members to teams you are not a member of."},
+                    status=400,
+                )
```

Comment:
would it be possible to break some of this logic down further into their own functions? we're doing a lot of stuff here and it would be helpful to isolate the logic for each check, e.g. all of the logic to check members inviting to member teams